src/basic/fileio.c

```diff
@@ -1354,3 +1354,44 @@ int link_tmpfile(int fd, const char *path, const char *target) {
 
         return 0;
 }
+
+int read_nul_string(FILE *f, char **ret) {
+        _cleanup_free_ char *x = NULL;
+        size_t allocated = 0, n = 0;
+
+        assert(f);
+        assert(ret);
+
+        /* Reads a NUL-terminated string from the specified file. */
+
+        for (;;) {
+                int c;
+
+                if (!GREEDY_REALLOC(x, allocated, n+2))
```

Comment:
This is a function call for every single byte that gets read. The whole function looks a bit cumbersome: Wouldn't it be easier to stop on NUL when writing the file, by just writing `strlen(buffer)` bytes? Then reading it could just slurp in entire blocks until EOF, i. e. this could re-use `read_full_file()`.